### DISCUSSION TITLE
Transactional create of three related models with two 1-n relationships between two of them

### DISCUSSION THREAD

**Discussion by nikolasburk** [prisma/prisma#5680] [Q&A]
I'm posting this as a follow up from a user request on [Slack](https://prisma.slack.com/archives/CCWDULGUW/p1613461766055800).

Here's the data model:

```prisma
model Company {
  company_id            Int         @id @default(autoincrement())
  name                  String      @unique
  users                 User[]
  workspacesAsInitiator Workspace[] @relation("InitiatedWorkspaces")
  workspacesAsResponder Workspace[] @relation("ResponderWorkspaces")
  @@map("companies")
}

model User {
  user_id    Int         @id @default(autoincrement())
  first_name String
  last_name  String?
  email      String      @unique
  password   String?
  role       user_role[]
  company_id Int?
  company    Company?    @relation(fields: [company_id], references: [company_id])
  @@map("users")
}

model Workspace {
  workspace_id Int     @id @default(autoincrement())
  initiator_id Int
  initiator    Company @relation("InitiatedWorkspaces", fields: [initiator_id], references: [company_id])
  responder_id Int
  responder    Company @relation("ResponderWorkspaces", fields: [responder_id], references: [company_id])
  @@map("workspaces")
}

enum user_role {
  CISO
  DPO
  Audit
  Security
  Owner
  Viewer
  Admin
}
```

The requirements the users has are as follows:

1. Create a `User` with the `Owner` role and a `Company` they belong to
2. Create more `User`s with `CISO` and `DPO` (and potentially even more) roles which belong to the same `Company`
3. Create a `Workspace` where both `responder` and `initiator` are the same `Company`

Is it possible to wrap this _entire_ flow inside of a transaction?

So far, I came up with the following flow:

```ts
  // Create owner with company
  const owner = await prisma.user.create({
    data: {
      first_name: 'Alice',
      email: 'alice@prisma.io',
      role: 'Owner',
      company: {
        create: {
          name: 'Prisma Corp.',
        },
      },
    },
    include: {
      company: true,
    },
  })

  // Create remaining users and workspace in a transaction
  if (owner.company) {
    const createUsers = prisma.user.createMany({
      data: [
        {
          first_name: 'Bob',
          email: 'bob@prisma.io',
          role: 'CISO',
          company_id: owner.company.company_id,
        },
        {
          first_name: 'Bob',
          email: 'bob@prisma.io',
          role: 'CISO',
          company_id: owner.company.company_id,
        },
        // you can add `Audit` and `Security` users here
      ],
    })
    const createWorkspace = prisma.workspace.create({
      data: {
        initiator_id: owner.company.company_id,
        responder_id: owner.company.company_id,
      },
    })
    await prisma.$transaction([
      createUsers,
      createWorkspace
    ])
  }
```

However, obviously the first `User` and `Company` will remain created even if part of the subsequent transaction fails. 

Ideally, this could be solved by a single nested write I guess because all models are related. But no matter from where you start the nested write, at some point it breaks down because of the two 1-n relationships between `Company` and `Workspace` 🤔

**Comment by nikolasburk** [comment]
After a quick internal discussion, one potential way to resolve this would be to use UUIDs or CUIDs instead of auto-incrementing integers as primary keys.

This is because right now the main issue is that the operation to create a workspace rely on the availability of the company ID that can only be generated by the DB.

If CUIDs or UUIDs were used, the ID could be generated by the application instead of the DB and could be provided inside a call to `$transaction`.

**Reply by nikolasburk** [reply]
Also note that this generally seems to be a good use case for [OCC](https://www.prisma.io/docs/guides/prisma-guides/prisma-client-transactions-guide#optimistic-concurrency-control) which we'll hopefully add to the Prisma Client API soon (you can follow this [issue](https://github.com/prisma/prisma/issues/4988) if you're interested in this.)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
